### ISSUE TITLE
[QIG-PURITY] WP4.2: Remove Euclidean Optimizers (Use Natural Gradient Only)

### ISSUE THREAD

**Issue by GaryOcean428** [GaryOcean428/pantheon-chat#76]
# Work Package 4.2: Remove Non-QIG Optimizers from Core Training

## Priority
**HIGH - GEOMETRIC CORRECTNESS**

## Objective
Ensure core QIG geometry training uses natural gradient / Fisher preconditioning, NOT standard Euclidean optimizers like Adam/SGD.

## Problem
Manifest explicitly forbids Adam/AdamW for "geometric learning" and prescribes natural gradient methods. Standard optimizers operate on Euclidean manifolds and will corrupt Fisher geometry.

## Current Risk
- Accidental use of `Adam()` in training loops
- Default PyTorch/TensorFlow optimizers (which are Euclidean)
- Copy-paste from tutorials that use standard optimization

### Natural Gradient Optimization
```python
# CORRECT - Fisher-aware optimization
optimizer = NaturalGradientOptimizer(
    parameters=model.parameters(),
    fisher_matrix=compute_fisher_information_matrix(),
    learning_rate=0.01
)
```

### Euclidean Optimizers (FOR COMPARISON ONLY - QUARANTINED)
```python
# WRONG in QIG-core, OK in baselines/
optimizer = torch.optim.Adam(model.parameters())  # Euclidean
```

### Audit Current Training Code
- [ ] Search for `Adam`, `AdamW`, `SGD`, `RMSprop` in QIG training code
- [ ] List all optimizer instantiations
- [ ] Document which are in QIG-core vs experiments

### Implement Fisher-Aware Optimizers
- [ ] Create `NaturalGradientOptimizer` class
- [ ] Implement Fisher matrix computation (or approximation)
- [ ] Add Fisher preconditioning to gradient updates
- [ ] Document approximations if exact Fisher is too expensive

### Replace Euclidean Optimizers
- [ ] Replace Adam/SGD with natural gradient in QIG-core
- [ ] Add validation: optimizer must be Fisher-aware
- [ ] Move any Euclidean optimizer usage to quarantine (baselines/)

### Add Validation
- [ ] Add check in training loop: `assert optimizer.is_fisher_aware()`
- [ ] Prevent accidental Euclidean optimization
- [ ] Log optimizer type at training start

### Documentation
- [ ] Document why natural gradient is required
- [ ] Explain Fisher preconditioning
- [ ] Add warning: "Do not use Adam in QIG-core"
- [ ] Link to papers on natural gradient methods

### Testing
- [ ] Create test: training with Adam should fail in QIG-core
- [ ] Create test: Fisher-aware optimizer should pass
- [ ] Verify gradient updates respect manifold

## Acceptance Criteria
- [ ] No Adam/SGD/RMSprop in QIG-core training
- [ ] All training uses Fisher-aware optimization
- [ ] Tests prevent accidental Euclidean optimization
- [ ] Clear documentation of optimizer requirements

## Context
From ChatGPT feedback: "Manifest forbids Adam/AdamW for geometric learning" - Euclidean optimizers will push parameters off the Fisher manifold.

## Coordination
⚠️ **DEPENDS ON**: #68 (canonical geometry module for Fisher computation)
⚠️ **CRITICAL**: This affects all training - coordinate with any ongoing training runs
⚠️ **NOTE**: May need to retrain existing models with correct optimizer

## References
- Type-Symbol-Concept Manifest: optimizer requirements
- Papers: Amari "Natural Gradient Works Efficiently in Learning"
- PyTorch geometry optimization libraries (geomstats, etc.)

**Comment by GaryOcean428** [comment]
## 🚨 INTEGRATION WITH MAMBA STATE SPACES

Natural gradient optimization is **even more critical** with Mamba (Granite 4.0-H) because Mamba's state space evolution IS natural gradient flow!

### Why Natural Gradient for QIG

**Euclidean gradient descent:**
```python
# Wrong: treats parameter space as flat
θ_new = θ_old - learning_rate * ∇L(θ)

# Problem: ignores manifold curvature
# Equivalent to: "Walk 1 meter north" works on flat earth, breaks on sphere
```

**Natural gradient descent:**
```python
# Correct: respects Fisher information geometry
θ_new = θ_old - learning_rate * F⁻¹ * ∇L(θ)
# where F = Fisher Information Matrix

# This is: "Walk along geodesic" (works on any manifold)
```

### Mamba State Spaces = Fisher Manifolds

**Critical insight from Granite 4.0-H:**

Mamba's state space model is:
```
dx/dt = Ax(t) + Bu(t)  # State evolution (SSM)
y(t) = Cx(t) + Du(t)   # Output
```

Fisher information flow is:
```
dx/dt = -∇_Fisher log p(x|θ)  # Natural gradient flow
```

**These are the SAME mathematical structure!** 

This means:
1. Mamba already operates on Fisher manifolds naturally
2. Natural gradient is the ONLY correct optimizer for Mamba-based models
3. Adam/SGD are geometrically incompatible (force flat-space assumptions)

### Implementation with Granite 4.0-H

```python
class NaturalGradientOptimizer:
    """
    Natural gradient optimization for QIG + Mamba.
    
    Uses Fisher Information Matrix preconditioning.
    Compatible with Granite 4.0-H state space structure.
    """
    
    def __init__(
        self,
        parameters,
        learning_rate: float = 0.01,
        fisher_approximation: str = "empirical"  # or "diagonal"
    ):
        self.params = parameters
        self.lr = learning_rate
        self.fisher_approx = fisher_approximation
        
    def compute_fisher_matrix(self, loss_fn, data_batch):
        """
        Compute Fisher Information Matrix.
        
        For Mamba models, this is NATIVE - we can extract it
        directly from the state space structure!
        """
        if self.fisher_approx == "empirical":
            # Compute from gradients of log-likelihood
            return self._empirical_fisher(loss_fn, data_batch)
        elif self.fisher_approx == "diagonal":
            # Diagonal approximation (faster)
            return self._diagonal_fisher(loss_fn, data_batch)
    
    def _empirical_fisher(self, loss_fn, data_batch):
        """
        Empirical Fisher via gradient outer products.
        
        F ≈ E[∇ log p(x|θ) ∇ log p(x|θ)ᵀ]
        """
        gradients = []
        
        for sample in data_batch:
            # Forward pass
            output = self.model(sample)
            log_prob = -loss_fn(output, sample)
            
            # Gradient of log-likelihood
            grad = torch.autograd.grad(log_prob, self.params)
            gradients.append(grad)
        
        # Outer product sum
        fisher = sum(
            torch.outer(g.flatten(), g.flatten())
            for g in gradients
        ) / len(data_batch)
        
        return fisher
    
    def _diagonal_fisher(self, loss_fn, data_batch):
        """
        Diagonal Fisher approximation (AdaGrad-style).
        
        Much faster, still respects geometry better than Adam.
        """
        squared_grads = []
        
        for sample in data_batch:
            output = self.model(sample)
            log_prob = -loss_fn(output, sample)
            grad = torch.autograd.grad(log_prob, self.params)
            squared_grads.append(grad ** 2)
        
        # Diagonal elements only
        fisher_diag = sum(squared_grads) / len(data_batch)
        return torch.diag(fisher_diag)
    
    def step(self, loss_fn, data_batch):
        """
        Natural gradient update step.
        
        θ_new = θ_old - lr * F⁻¹ * ∇L
        """
        # Compute Fisher matrix
        F = self.compute_fisher_matrix(loss_fn, data_batch)
        
        # Add damping for numerical stability
        F_damped = F + 1e-4 * torch.eye(F.shape[0])
        
        # Compute loss gradient
        loss = loss_fn(data_batch)
        grad_loss = torch.autograd.grad(loss, self.params)
        
        # Natural gradient: F⁻¹ * grad
        natural_grad = torch.linalg.solve(F_damped, grad_loss)
        
        # Update parameters
        for param, nat_grad in zip(self.params, natural_grad):
            param.data -= self.lr * nat_grad
```

### For Mamba Models: Direct State Space Access

```python
class MambaNaturalGradient(NaturalGradientOptimizer):
    """
    Specialized for Granite 4.0-H Mamba models.
    
    Exploits state space structure for efficient Fisher computation.
    """
    
    def compute_fisher_from_mamba_state(self, mamba_layer):
        """
        Extract Fisher matrix directly from Mamba state space.
        
        Mamba: dx/dt = Ax + Bu
        Fisher metric: F = ∂x/∂θ · ∂x/∂θᵀ
        
        We can compute this from A, B, C, D matrices!
        """
        A = mamba_layer.A  # State transition
        B = mamba_layer.B  # Input mapping
        C = mamba_layer.C  # Output mapping
        
        # Fisher metric from state space Jacobian
        # This is EXACT for SSMs, not approximate!
        fisher = self._compute_ssm_fisher(A, B, C)
        
        return fisher
```

### Validation Against Adam (For Science)

```python
# Quarantine: docs/08-experiments/baselines/adam_comparison.py

def compare_optimizers():
    """
    Compare natural gradient vs Adam (for validation only).
    
    Expected: Natural gradient converges faster and to higher Φ.
    """
    
    results = {
        "natural_gradient": train_with_natural_gradient(),
        "adam": train_with_adam(),  # Euclidean (wrong geometry)
    }
    
    print(f"Final Φ (Natural): {results['natural_gradient']['phi']}")
    print(f"Final Φ (Adam):    {results['adam']['phi']}")
    print(f"Convergence (NG):  {results['natural_gradient']['epochs']}")
    print(f"Convergence (Adam): {results['adam']['epochs']}")
```

### Acceptance Criteria (Updated)

- [ ] NaturalGradientOptimizer class implemented
- [ ] Empirical Fisher computation (gradient outer products)
- [ ] Diagonal Fisher approximation (faster variant)
- [ ] MambaNaturalGradient subclass for Granite 4.0-H
- [ ] Direct state space Fisher extraction
- [ ] Training loop uses natural gradient only (no Adam/SGD)
- [ ] Validation test (quarantined): natural gradient > Adam
- [ ] Integration with #68 (canonical geometry module)
- [ ] Works with #76 (foresight/waypoint planning)

### Why This Matters

**Adam on Mamba = Forcing flat geometry on curved manifold**
- Like using Euclidean distance on a sphere
- Breaks the native geometric structure
- Corrupts consciousness emergence

**Natural gradient on Mamba = Native geometric flow**
- Respects the manifold curvature
- Preserves state space structure
- Enables true geometric learning

**With Granite 4.0-H, we can extract Fisher matrix directly from the model architecture - this is a HUGE advantage!** 🌊✨
